Add SPI to WKWebView for toggling in-window mode

danae404 opened the pull request:
#### 7d3c8dd214517a4bda7aeb8aa5f609d16b74631c
<pre>
Add SPI to WKWebView for toggling in-window mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=269032">https://bugs.webkit.org/show_bug.cgi?id=269032</a>
<a href="https://rdar.apple.com/122593920">rdar://122593920</a>

Reviewed by Jer Noble.

This patch adds SPI to WKWebView _canToggleInWindow and _toggleInWindow, and a stub
for _isInWindowActive.

_canToggleInWindow uses the same heuristic as _canTogglePictureInPicture
To determine if there is a viable video on the webpage to put into in-window
Fullscreen mode.

_toggleInWindow toggles a viable video, if there is one, between
In-window and inline mode.

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::toggleInWindow):
(WebCore::PlaybackSessionModelMediaElement::isInWindowActive):
* Source/WebCore/platform/mac/WebPlaybackControlsManager.h:
* Source/WebCore/platform/mac/WebPlaybackControlsManager.mm:
(-[WebPlaybackControlsManager toggleInWindow]):
(-[WebPlaybackControlsManager isInWindowActive]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _canToggleInWindow]):
(-[WKWebView _isInWindowActive]):
(-[WKWebView _toggleInWindow]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::toggleInWindow):
(WebKit::PlaybackSessionModelContext::isInWindowActive):
(WebKit::PlaybackSessionManagerProxy::toggleInWindow):
(WebKit::PlaybackSessionManagerProxy::isInWindowActive):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::isInWindowActive):
(WebKit::WebViewImpl::toggleInWindow):
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionManager::toggleInWindow):
(WebKit::PlaybackSessionManager::isInWindowActive):

Canonical link: <a href="https://commits.webkit.org/274650@main">https://commits.webkit.org/274650@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cdaf89f6d19b2159ea2b39105428da77fa137616

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39222 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18201 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41756 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35122 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41528 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21054 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15530 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32826 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39796 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15325 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34004 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13309 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13285 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34943 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43034 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/32675 "Build is in progress. Recent messages:") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35620 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35273 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39093 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14034 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11589 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37405 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15640 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/8852 "The change is no longer eligible for processing.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15303 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5206 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15126 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2539,6 +2539,15 @@ - (BOOL)_canTogglePictureInPicture
 #endif
 }
 
+- (BOOL)_canToggleInWindow
+{
+#if HAVE(TOUCH_BAR)
+    return _impl->canTogglePictureInPicture();
+#else
+    return NO;
+#endif
+}
+
```
This is fine for now, but I think we will eventually want a non HAVE_TOUCH_BAR path.

jernoble commented on Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
```diff
@@ -39,6 +39,8 @@ messages -> PlaybackSessionManager {
     HandleControlledElementIDRequest(WebKit::PlaybackSessionContextIdentifier contextId)
     TogglePictureInPicture(WebKit::PlaybackSessionContextIdentifier contextId)
+    ToggleInWindow(WebKit::PlaybackSessionContextIdentifier contextId)
+    IsInWindowActive(WebKit::PlaybackSessionContextIdentifier contextId)
```
This should change to a push message from the WebContent process to the UIProcess via the PlaybackSessionManagerProxy. When the WebContent process detects a change it in-window mode, it should send a message up.

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -367,9 +367,12 @@ for this property.
 - (void)_stopMediaCapture WK_API_AVAILABLE(macos(10.15.4), ios(13.4));
 
 @property (nonatomic, readonly) BOOL _canTogglePictureInPicture;
+@property (nonatomic, readonly) BOOL _canToggleInWindow;
```
don’t these need `WK_API_AVAILABLE`?

mahruskazi commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2548,6 +2557,11 @@ - (BOOL)_isPictureInPictureActive
 
+- (BOOL)_isInWindowActive
+{
+    return NO;
```
Are we going to fix this in a later patch?

danae404 replied:
Yes!

danae404 committed:
Add SPI to WKWebView for toggling in-window mode
https://bugs.webkit.org/show_bug.cgi?id=269032
rdar://122593920

Reviewed by Jer Noble.

This patch adds SPI to WKWebView _canToggleInWindow and _toggleInWindow, and a stub
for _isInWindowActive.

_canToggleInWindow uses the same heuristic as _canTogglePictureInPicture
To determine if there is a viable video on the webpage to put into in-window
Fullscreen mode.

_toggleInWindow toggles a viable video, if there is one, between
In-window and inline mode.

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::toggleInWindow):
(WebCore::PlaybackSessionModelMediaElement::isInWindowActive):
* Source/WebCore/platform/mac/WebPlaybackControlsManager.h:
* Source/WebCore/platform/mac/WebPlaybackControlsManager.mm:
(-[WebPlaybackControlsManager toggleInWindow]):
(-[WebPlaybackControlsManager isInWindowActive]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _canToggleInWindow]):
(-[WKWebView _isInWindowActive]):
(-[WKWebView _toggleInWindow]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::toggleInWindow):
(WebKit::PlaybackSessionModelContext::isInWindowActive):
(WebKit::PlaybackSessionManagerProxy::toggleInWindow):
(WebKit::PlaybackSessionManagerProxy::isInWindowActive):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::isInWindowActive):
(WebKit::WebViewImpl::toggleInWindow):
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionManager::toggleInWindow):
(WebKit::PlaybackSessionManager::isInWindowActive):

Canonical link: https://commits.webkit.org/274650@main
